xpartition.py

```diff
@@ -187,13 +189,44 @@ def _write_partition_dataarray(
         ds.isel(partition).to_zarr(store, region=partition)
 
 
-def _write_partition_dataset(
-    ds: xr.Dataset, store: str, ranks: int, dims: Sequence[Hashable], rank: int
-):
+@dataclass
+class HashableIndexers:
+    indexers: Region
+
+    @property
+    def immutable_representation(self):
+        if self.indexers is None:
+            return self.indexers
+        indexers = ((k, (s.start, s.stop, s.step)) for k, s in self.indexers.items())
+        return tuple(sorted(indexers, key=lambda x: x[0]))
+
+    def __hash__(self):
+        return hash(self.immutable_representation)
+
+
+def _collect_by_partition(
+    ds: xr.Dataset, ranks: int, dims: Sequence[Hashable], rank: int
+) -> Sequence[Tuple[Region, xr.Dataset]]:
+    """Return a list of pairs of partitions and Datasets containing
+    DataArrays that can be written out to those partitions.
+    """
+    dataarrays = collections.defaultdict(list)
     for da in ds.data_vars.values():
         if isinstance(da.data, dask.array.Array):
             partition_dims = [dim for dim in dims if dim in da.dims]
-            da.partition.write(store, ranks, partition_dims, rank)
+            indexers = da.partition.indexers(ranks, rank, partition_dims)
+            key = HashableIndexers(indexers)
+            dataarrays[key].append(da)
+    return [(k.indexers, xr.merge(v)) for k, v in dataarrays.items()]
```

Comment:
Maybe use a function instead of the class `HashableIndexers`, which is a bit of an odd concept since it contains mutable state but advertises itself as a constant. Maybe it's a small risk, but the indexers could be modified after they are placed in the `dataarrays` dict...this would be impossible if you explicitly converted between the immutable and mutable representations.
```suggestion
            dataarrays[freeze(indexers)].append(da)
    return [(unfreeze(frozen_indexer), xr.merge(v)) for frozen_indexer, v in dataarrays.items()]
```

Reply:
Very true -- using functions is a cleaner design.